src/analysis/complex/upper_half_plane.lean

```diff
@@ -0,0 +1,169 @@
+/-
+Copyright (c) 2021 Alex Kontorovich and Heather Macbeth and Marc Masdeu. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Kontorovich, Heather Macbeth, Marc Masdeu
+-/
+
+import linear_algebra.special_linear_group
+import analysis.complex.basic
+import group_theory.group_action.defs
+
+/-!
+# The upper half plane and its automorphisms
+
+This file defines `upper_half_plane` to be the upper half plane in `ℂ`.
+
+We furthermore equip it with the structure of an `SL(2,ℝ)` action by
+fractional linear transformations.
+
+We define the notation `ℍ` for the upper half plane available in the locale
+`upper_half_plane` so as not to conflict with the quaternions.
+-/
+
+noncomputable theory
+
+open matrix matrix.special_linear_group
+
+open_locale classical big_operators
+
+local attribute [instance] fintype.card_fin_even
+
+/-- The upper half plane -/
+abbreviation upper_half_plane :=
+{point : ℂ // 0 < point.im}
+
+localized "notation `ℍ` := upper_half_plane" in upper_half_plane
+
+namespace upper_half_plane
+
+/-- Imaginary part -/
+def im (z : ℍ) := z.1.im
+
+/-- Real part -/
+def re (z : ℍ) := z.1.re
+
+@[simp] lemma coe_point (z : ℍ) : z.1 = (z:ℂ) := rfl
+
+@[simp] lemma coe_im (z : ℍ) : (z:ℂ).im = z.im := rfl
+
+@[simp] lemma coe_re (z : ℍ) : (z:ℂ).re = z.re := rfl
+
+lemma im_pos (z : ℍ) : 0 < z.im := z.2
+
+lemma im_ne_zero (z : ℍ) : z.im ≠ 0 := z.im_pos.ne'
+
+lemma ne_zero (z : ℍ) : (z : ℂ) ≠ 0 :=
+begin
+  intros h,
+  apply z.im_ne_zero,
+  simp [← complex.zero_im, ← h]
+end
+
+lemma norm_sq_pos (z: ℍ) : 0 < complex.norm_sq (z : ℂ) :=
+begin
+  rw complex.norm_sq_pos,
+  exact z.ne_zero
+end
+
+lemma norm_sq_ne_zero (z : ℍ) : complex.norm_sq (z : ℂ) ≠ 0 := ne_of_gt (norm_sq_pos z)
+
+local notation `SL(` n `,` R `)`:= special_linear_group (fin n) R
```

Comment:
Should this move to just after the definition of `special_linear_group`, and become localized notation?
And then in this file something like `open_locale matrix_groups`?